### PR TITLE
[Vengeance] add base T30 4pc implementation

### PR DESCRIPTION
**Pull request by ToppleTheNun**
T29 damage breakdown:
image

T30 damage breakdown:
image

### CODE REVIEW

**Comment by ToppleTheNun** [comment]
@EvanMichaels

File: engine/class_modules/sc_demon_hunter.cpp [hunk 1]
```diff
@@ -4717,6 +4721,14 @@ struct fracture_t : public demon_hunter_attack_t
       mh->set_target( s->target );
       oh->set_target( s->target );
       mh->execute();
+      // t30 4pc proc happens on main hand execute but before off hand execute
+      // this matters because the off hand hit benefits from Fiery Demise that may be
+      // applied because of the t30 4pc proc
+      if ( p()->buff.t30_vengeance_4pc->up() )
+      {
+        p()->active.fiery_brand_t30->execute_on_target( s->target );
+        p()->buff.t30_vengeance_4pc->reset();
```
Comment (EvanMichaels) [inline engine/class_modules/sc_demon_hunter.cpp]: ->reset() is for end of iteration and resets some other things like last_start, last_trigger, etc. Should just use ->expire() here.

File: engine/class_modules/sc_demon_hunter.cpp [hunk 2]
```diff
@@ -4786,6 +4798,12 @@ struct shear_t : public demon_hunter_attack_t
         p()->proc.soul_fragment_from_meta->occur();
       }
     }
+
+    if ( p()->buff.t30_vengeance_4pc->up() )
+    {
+      p()->active.fiery_brand_t30->execute_on_target( s->target );
+      p()->buff.t30_vengeance_4pc->reset();
```
Comment (EvanMichaels) [inline engine/class_modules/sc_demon_hunter.cpp]: ->reset() is for end of iteration and resets some other things like last_start, last_trigger, etc. Should just use ->expire() here.

File: engine/class_modules/sc_demon_hunter.cpp [hunk 3]
```diff
@@ -553,6 +553,7 @@ class demon_hunter_t : public player_t
     double t30_havoc_2pc_fury_tracker = 0.0;
     const spell_data_t* t30_vengeance_2pc_buff;
     const spell_data_t* t30_vengeance_4pc_buff;
+    double t30_vengeance_4pc_soul_fragments_tracker = 0;
```
Comment (EvanMichaels) [inline engine/class_modules/sc_demon_hunter.cpp:556]: You will need to add this to demon_hunter_t::reset() to set it back to 0 on iteration reset

**Review by EvanMichaels — CHANGES_REQUESTED** [review]

**Review by EvanMichaels — APPROVED** [review]

**Pull request merged**